File: mobile/src/main/java/info/papdt/express/helper/ui/shortcut/ScannerShortcutCreator.java
package info.papdt.express.helper.ui.shortcut;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.os.Parcelable;

import info.papdt.express.helper.R;
import info.papdt.express.helper.ui.MainActivity;
import info.papdt.express.helper.ui.ScannerActivity;

public class ScannerShortcutCreator extends Activity {

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		if (getIntent().getAction().equals(Intent.ACTION_CREATE_SHORTCUT)) {
			Intent intent = new Intent();
			Parcelable icon = Intent.ShortcutIconResource.fromContext(this, R.mipmap.ic_scanner_launcher);

			intent.putExtra(Intent.EXTRA_SHORTCUT_NAME, getString(R.string.activity_scanner));
			intent.putExtra(Intent.EXTRA_SHORTCUT_ICON_RESOURCE, icon);

			Intent launchIntent = new Intent(getApplicationContext(), MainActivity.class);
			launchIntent.setAction(ScannerActivity.ACTION_SCAN_TO_ADD);
			launchIntent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_NEW_TASK);

			intent.putExtra(Intent.EXTRA_SHORTCUT_INTENT, launchIntent);

			setResult(RESULT_OK, intent);
		} else {
			setResult(RESULT_CANCELED);
		}
		finish();
	}

}

File: mobile/src/main/kotlin/info/papdt/express/helper/ui/shortcut/ScannerShortcutCreator.kt
package info.papdt.express.helper.ui.shortcut

import android.app.Activity
import android.content.ComponentName
import android.content.Intent
import android.os.Bundle
import androidx.core.content.pm.ShortcutInfoCompat
import androidx.core.content.pm.ShortcutManagerCompat
import androidx.core.graphics.drawable.IconCompat

import info.papdt.express.helper.R
import info.papdt.express.helper.ui.HomeActivity
import info.papdt.express.helper.ui.ScannerActivity

class ScannerShortcutCreator : Activity() {

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		if (intent.action == Intent.ACTION_CREATE_SHORTCUT) {
			val icon = IconCompat.createWithResource(this, R.mipmap.ic_scanner_launcher)
			val intent = ShortcutManagerCompat.createShortcutResultIntent(
					this,
					ShortcutInfoCompat.Builder(this, "scanner")
							.setActivity(ComponentName(this, HomeActivity::class.java))
							.setIntent(Intent(ScannerActivity.ACTION_SCAN_TO_ADD)
									.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
									.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK))
							.setIcon(icon)
							.setShortLabel(getString(R.string.activity_scanner))
							.setLongLabel(getString(R.string.activity_scanner))
							.build()
			)
			setResult(Activity.RESULT_OK, intent)
		} else {
			setResult(Activity.RESULT_CANCELED)
		}
		finish()
	}

}
